lib/commons/aria/roles.js

```diff
@@ -204,3 +202,252 @@ aria.implicitRole = function(node) {
+			if (node.alt) {
+				return !isApplicable;
+			}
+			return isApplicable;
+		},
+		INPUT: ({ node, role, isApplicable }) => {
+			switch (node.type) {
+				case 'button':
+				case 'image':
+					return isApplicable;
+				case 'checkbox':
+					if (role === 'button' && node.hasAttribute('aria-pressed')) {
+						return true;
+					}
+					return isApplicable;
+				case 'radio':
+					return role === 'menuitemradio';
+				default:
+					return false;
+			}
+		},
+		LI: ({ node, isApplicable }) => {
+			const hasImplicitListitemRole = node.matches('ol li, ul li');
+			if (hasImplicitListitemRole) {
+				return isApplicable;
+			}
+			return true;
+		},
+		LINK: ({ node }) => !node.href,
+		MENU: ({ node }) => {
+			if (node.getAttribute('type') === 'context') {
+				return false;
+			}
+			return true;
+		},
+		OPTION: ({ node }) => {
+			const withinOptionList = node.matches(
+				'select > option, datalist > option, optgroup > option'
+			);
+			return !withinOptionList;
+		},
+		SELECT: ({ node, role }) =>
+			!node.multiple && node.size <= 1 && role === 'menu',
+		SVG: ({ node, isApplicable }) => {
+			// if in svg context it all roles may be used
+			if (
+				node.parentNode &&
+				node.parentNode.namespaceURI === 'http://www.w3.org/2000/svg'
+			) {
+				return true;
+			}
+			return isApplicable;
+		}
+	};
+
+	const lookupTableRole = axe.commons.aria.lookupTable.role;
+	const canHaveNoRole = lookupTableRole.NO_ROLE.allowedElements.includes(
+		tagName
+	);
+	if (canHaveNoRole) {
+		return false;
+	}
+	const canHaveAnyRole = lookupTableRole.ANY_ROLE.allowedElements.includes(
+		tagName
+	);
+	if (canHaveAnyRole) {
+		return true;
+	}
+
+	let isApplicable = false;
+	if (lookupTableRole[role]) {
+		isApplicable = isApplicableRoleForNode(node, lookupTableRole[role]);
+	}
+	if (elementCheck[tagName]) {
+		// check if tag has a custom function to check its validity against various tags
+		isApplicable = elementCheck[tagName]({ node, role, isApplicable });
+	}
+	return isApplicable;
+}
+
+/**
+ * gets all unallowed roles for a given node
+ * @method getAriaUnAllowedRoles
+ * @param {Object} node HTMLElement to validate
+ * @param {String} tagName tag name of a node
+ * @param {String} allowImplicit option to allow implicit roles, defaults to true
+ * @return {Array<String>} retruns an array of roles that are not allowed on the given node
+ */
+function getAriaUnAllowedRoles(node, tagName, allowImplicit) {
+	/**
+	 * Get roles applied to a given node
+	 * @param {HTMLElement} node HTMLElement
+	 * @return {Array<String>} return an array of roles applied to the node, if no roles, return an empty array.
+	 */
+	function getRoleSegments(node) {
+		let roles = [];
+		if (!node) {
+			return roles;
+		}
+		if (node.hasAttribute('role')) {
+			const nodeRoles = axe.utils.tokenList(
+				node.getAttribute('role').toLowerCase()
+			);
+			roles = roles.concat(nodeRoles);
+		}
+		if (node.hasAttributeNS('http://www.idpf.org/2007/ops', 'type')) {
```

Comment:
`epub:type` shouldn't be looked up in this function, as it has absolutely no accessibility support and most probably never will. I'd suggest removing this whole `if` block.

Reply:
That's very good to know. Thanks Romain.